src/ert/storage/local_ensemble.py

```diff
@@ -876,3 +900,127 @@ def get_response_state(
+                }
+
+                reals = iens_active_index.tolist()
+                reals.sort()
+                # too much memory to do it all at once, go per realization
+                first_columns: polars.DataFrame | None = None
+                realization_columns: List[polars.DataFrame] = []
+                for real in reals:
+                    responses = self._load_responses_lazy(
+                        response_type, (real,)
+                    ).with_columns(
+                        [
+                            polars.col("response_key")
+                            .cast(polars.Categorical)
+                            .alias("response_key")
+                        ]
+                    )
+
+                    # Filter out responses without observations
+                    for col, observed_values in observed_cols.items():
+                        responses = responses.filter(
+                            polars.col(col).is_in(observed_values)
+                        )
+
+                    pivoted = responses.collect().pivot(
+                        on="realization",
+                        index=["response_key", *response_cls.primary_key],
+                        values="values",
+                        aggregate_function="mean",
+                    )
+
+                    if pivoted.is_empty():
+                        joined = observations_for_type.with_columns(
+                            polars.Series(
+                                str(real),
+                                [np.nan] * len(observations_for_type),
+                                dtype=polars.Float32,
+                            )
+                        )
+                    elif "time" in pivoted:
+                        joined = observations_for_type.join_asof(
+                            pivoted,
+                            by=["response_key", *response_cls.primary_key],
+                            on="time",
+                            tolerance="1s",
+                        )
+                    else:
+                        joined = observations_for_type.join(
+                            pivoted,
+                            how="left",
+                            on=["response_key", *response_cls.primary_key],
+                        )
+
+                    joined = (
+                        joined.with_columns(
+                            polars.concat_str(
+                                response_cls.primary_key, separator=", "
+                            ).alias(
+                                "__tmp_index_key__"
+                                # Avoid potential collisions w/ primary key
+                            )
+                        )
+                        .drop(response_cls.primary_key)
+                        .rename({"__tmp_index_key__": "index"})
+                    )
+
+                    # if not joined.is_empty():
+                    if first_columns is None:
```

Comment:
When can this happen?

Reply:
It will happen in the first iteration through this loop. first_columns is the `index` columns, and they will be aligned for all realizations, so we only need to extract it once.

Reply:
Ahh, that was not apparent to me, suggest making more explicit, either by enumerating the loop or adding a comment, was a bit confusing with two seemingly related checks.